numba/tests/test_dispatcher.py

```diff
@@ -1746,5 +1750,103 @@ def foo():
         self.assertRegexpMatches(str(raises.exception), pat)
 
 
+class TestNoRetryFailedSignature(unittest.TestCase):
+    """Test that failed-to-compile signatures are not recompiled.
+    """
+    def test_direct_call(self):
+        @jit(nopython=True)
+        def foo(x):
+            return x[0]
+
+        fcom = foo._compiler
+        self.assertEqual(len(fcom._failed_cache), 0)
+        # expected failure because `int` has no `__getitem__`
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry with double
+        with self.assertRaises(errors.TypingError):
+            foo(1.0)
+        self.assertEqual(len(fcom._failed_cache), 2)
+
+    def test_nested_call(self):
+        @jit(nopython=True)
+        def bar(x):
+            return x[0]
+
+        @jit(nopython=True)
+        def foobar(x):
+            bar(x)
+
+        @jit(nopython=True)
+        def foo(x):
+            return bar(x) + foobar(x)
+
+        fcom = foo._compiler
+        self.assertEqual(len(fcom._failed_cache), 0)
```

Comment:
the following ~20 lines of asserting things is a repeat of the above? fold into a function and call it?